swarm/animator.py

```diff
@@ -45,3 +59,306 @@ def animate(i):
+            FuncAnimation used here. These are created in the `init` and
+            used in the animate.
+    """
+
+    artists = attr.ib(init=False, factory=list, type=List[plt.Artist])
+
+    @property
+    def num_frames(self) -> int:
+        """Nice to have, allows animation to know how long to go for"""
+        return 0
+
+    def init(self, ax: plt.Axes):
+        """Populate the artists and call the hooks"""
+        pass
+
+    def animate(self, frame: int):
+        """Move the artists"""
+        pass
+
+
+def _validate_ax_kwargs(**kwargs):
+    """
+    This is an attempt to validate the kwargs on definition rather than at runtime for quicker
+    feedback to the developer.
+    This isn't needed by end-user, but could be used ahead of time if they like.
+
+    Raises:
+        ValueError: in the case of invalid arguments
+    """
+    cls = plt.Axes
+    missing = []
+    for k in kwargs:
+        if not hasattr(cls, k):
+            missing.append(k)
+    if missing:
+        raise ValueError("plt.Axes does not have {}".format(",".join(missing)))
+
+
+def _kwargs_hook(**kwargs):
+    """
+    Our animation library uses hooks to allow once of configuration of an axes. These
+    include things like plotting the true y value or setting sensible x/y limits for the plot
+
+    This can be time consuming to do, so this provides an easy way to do so, allowing
+    one to call the various functions of ``plt.Axes`` as below. This is expected to be fit
+    into the existing machinery of the SwarmPlot sub/duck classes so you won't need to make
+    this call directly
+
+    Examples:
+        >>> _kwargs_hook(set_title="Goofy Experiment", set_ylabel="loss")
+        # will be used mostly with constructors like
+        >>> LineSwarm.standard(x, y, data, set_title="Goofy Experiment", set_ylabel="loss")
+
+    Raises:
+        ValueError: will attempt to validate the kwargs ahead of time, but also on execution
+        in case the function arg is incorrect
+
+    """
+    # get an early error if this is a problem
+    # rather than at animate time
+    _validate_ax_kwargs(**kwargs)
+
+    def inner_hook(ax):
+        for k, v in kwargs.items():
+            try:
+                func = getattr(ax, k)
+            except AttributeError:
+                raise ValueError(f"Used a method {k}")
+            else:
+                try:
+                    func(v)
+                except Exception as e:
+                    # we might have an issue that we're hitting a property
+                    # or have type mismatches
+                    raise ValueError("Unexpected failure") from e
+
+    return inner_hook
+
+
+def _apply_hook(ax: plt.Axes, hook: Union[Iterable[Callable], Callable, None]):
+    """Applies a list of hooks to the axes object"""
+    if not hook:
+        return
+    try:
+        for h in hook:
+            h(ax)
+    except TypeError:
+        # can't iterate, must be single callabe
+        hook(ax)
```

Comment:
So the hooks here are not very well defined, explained or designed so this can be confusing.

But the basic idea is that you want to pass a function to act on some kind of object to another function. For example `summarise_across(data, np.mean)` is a very simple form of a hook. 
Hooks are commonly found in classes, not functions, but they're not always necessary. For example, if you wanted to customise how `LineSwarm` did it's axes init you could instead do this:

```
class MyLineSwarm(LineSwarm):
    def hook(self, ax):
        # do stuff
```

Hooks tend to be a way of reducing code replication here and multiple single classes, and by having separate hooks, you can combine them in powerful ways (say be having filters and in place modification).

Reply:
Ok, cool, thanks!  I think I'll try to read a few more examples somewhere.  Seems quite powerful.